cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java

```diff
@@ -402,6 +402,23 @@ private void handleProgramEvent(ProgramRunId programRunId, ProgramRunStatus prog
         recordedRunRecord = appMetadataStore.recordProgramResumed(programRunId, messageIdBytes, resumeTime);
         writeToHeartBeatTable(recordedRunRecord, resumeTime, programHeartbeatTable);
         break;
+      case STOPPING:
+        Map<String, String> notificationProperties = notification.getProperties();
+        long stoppingTimestamp = getTimeSeconds(notificationProperties, ProgramOptionConstants.STOPPING_TIME);
+        if (stoppingTimestamp == -1) {
+          LOG.warn("Ignore program stopping notification for program {} without {} specified, {}",
+                   programRunId, ProgramOptionConstants.STOPPING_TIME, notification);
+          return;
+        }
+        String gracefulShutdownPeriod = notificationProperties.get(
+          ProgramOptionConstants.GRACEFUL_SHUTDOWN_PERIOD_IN_SECS);
+        int gracefulShutdownSecs = (gracefulShutdownPeriod == null) ? Integer.MAX_VALUE :
+          Integer.parseInt(gracefulShutdownPeriod);
+        long terminateTimestamp = stoppingTimestamp + gracefulShutdownSecs;
```

Comment:
When calling something timestamp, always use millis. It is better to store millis in run record instead of seconds.